### ISSUE TITLE
NPE when PARTITION BY derived field on null-valued stream record

### ISSUE THREAD

**Issue by mikebin** [confluentinc/ksql#6687]
**Describe the bug**
An NPE occurs in the Kafka Streams layer when executing a `PARTITION BY` query on a derived field (e.g., the result of a function call) on a null-valued stream record. This causes the Streams app to fail, and the query to go to `ERROR` state.

**To Reproduce**
Versions: 6.0.0, 0.13.0

1. Create a simple stream:
```sql
create stream s (f1 string, f2 string) with (kafka_topic='s', value_format='json', partitions=1);
```

2. Produce a null-valued record to the source topic (for example, with `kafkacat`):
```
kafkacat -P -t s  -b localhost:9092 -K"#" -Z << EOF
k1#
EOF
```

3. Create CSAS which partitions by a derived field from stream `s`:
```sql
create stream npe as select * from s partition by ucase(f2);
```

**Expected behavior**
ksqlDB will silently drop null-valued record, or otherwise handle the exception and continue executing the query.

**Actual behavior**
Query goes to `ERROR` state (in ksqlDB 0.13.0, it'll be automatically restarted, but will continue to fail):

```
ksql> show queries;

 Query ID   | Query Type | Status  | Sink Name | Sink Kafka Topic | Query String
----------------------------------------------------------------------------------------------------------------------------------------------------------------
 CSAS_NPE_0 | PERSISTENT | ERROR:1 | NPE       | NPE              | CREATE STREAM NPE WITH (KAFKA_TOPIC='NPE', PARTITIONS=1, REPLICAS=1) AS SELECT * FROM S S PARTITION BY UCASE(S.F2) EMIT CHANGES;
----------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Exceptions in the log:

```
[2020-11-27 23:07:37,367] ERROR {"type":1,"deserializationError":null,"recordProcessingError":{"errorMessage":"Error computing new key from expression UCASE(F2)","record":null,"cause":["java.lang.NullPointerException"]},"productionError":null} (processing.CSAS_NPE_0.PartitionBy:44)
[2020-11-27 23:07:37,367] ERROR {"type":1,"deserializationError":null,"recordProcessingError":{"errorMessage":"Error computing new key from expression UCASE(F2)","record":null,"cause":["java.lang.NullPointerException"]},"productionError":null} (processing.CSAS_NPE_0.PartitionBy:44)
[2020-11-27 23:07:37,368] ERROR stream-thread [_confluent-ksql-default_query_CSAS_NPE_0-e42dd5be-29f4-41ca-a547-91404392c87a-StreamThread-1] Failed to process stream task 0_0 due to the following error: (org.apache.kafka.streams.processor.internals.TaskManager:1044)
org.apache.kafka.streams.errors.StreamsException: Exception caught in process. taskId=0_0, processor=KSTREAM-SOURCE-0000000000, topic=s, partition=0, offset=0, stacktrace=java.lang.NullPointerException
	at io.confluent.ksql.execution.streams.PartitionByParamsFactory.lambda$buildMapper$2(PartitionByParamsFactory.java:166)
	at org.apache.kafka.streams.kstream.internals.KStreamMap$KStreamMapProcessor.process(KStreamMap.java:41)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.kstream.internals.KStreamTransformValues$KStreamTransformValuesProcessor.process(KStreamTransformValues.java:64)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:96)
	at org.apache.kafka.streams.processor.internals.StreamTask.lambda$process$1(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.TaskManager.process(TaskManager.java:1033)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:690)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:551)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:510)

	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:696)
	at org.apache.kafka.streams.processor.internals.TaskManager.process(TaskManager.java:1033)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:690)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:551)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:510)
Caused by: java.lang.NullPointerException
	at io.confluent.ksql.execution.streams.PartitionByParamsFactory.lambda$buildMapper$2(PartitionByParamsFactory.java:166)
	at org.apache.kafka.streams.kstream.internals.KStreamMap$KStreamMapProcessor.process(KStreamMap.java:41)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.kstream.internals.KStreamTransformValues$KStreamTransformValuesProcessor.process(KStreamTransformValues.java:64)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:96)
	at org.apache.kafka.streams.processor.internals.StreamTask.lambda$process$1(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:679)
	... 4 more
[2020-11-27 23:07:37,370] ERROR stream-thread [_confluent-ksql-default_query_CSAS_NPE_0-e42dd5be-29f4-41ca-a547-91404392c87a-StreamThread-1] Encountered the following exception during processing and the thread is going to shut down:  (org.apache.kafka.streams.processor.internals.StreamThread:530)
org.apache.kafka.streams.errors.StreamsException: Exception caught in process. taskId=0_0, processor=KSTREAM-SOURCE-0000000000, topic=s, partition=0, offset=0, stacktrace=java.lang.NullPointerException
	at io.confluent.ksql.execution.streams.PartitionByParamsFactory.lambda$buildMapper$2(PartitionByParamsFactory.java:166)
	at org.apache.kafka.streams.kstream.internals.KStreamMap$KStreamMapProcessor.process(KStreamMap.java:41)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.kstream.internals.KStreamTransformValues$KStreamTransformValuesProcessor.process(KStreamTransformValues.java:64)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:96)
	at org.apache.kafka.streams.processor.internals.StreamTask.lambda$process$1(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.TaskManager.process(TaskManager.java:1033)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:690)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:551)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:510)

	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:696)
	at org.apache.kafka.streams.processor.internals.TaskManager.process(TaskManager.java:1033)
	at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:690)
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:551)
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:510)
Caused by: java.lang.NullPointerException
	at io.confluent.ksql.execution.streams.PartitionByParamsFactory.lambda$buildMapper$2(PartitionByParamsFactory.java:166)
	at org.apache.kafka.streams.kstream.internals.KStreamMap$KStreamMapProcessor.process(KStreamMap.java:41)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.kstream.internals.KStreamTransformValues$KStreamTransformValuesProcessor.process(KStreamTransformValues.java:64)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.lambda$process$2(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:142)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:236)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:216)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:168)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:96)
	at org.apache.kafka.streams.processor.internals.StreamTask.lambda$process$1(StreamTask.java:679)
	at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImpl.maybeMeasureLatency(StreamsMetricsImpl.java:836)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:679)
	... 4 more
```

**Additional context**

**Comment by spena** [comment]
Will discuss with the team to mark this as P0 or P1 for 0.15 as it is causing service disruption when the null value is found.

**Comment by jzaralim** [comment]
There is a fix for the NPE here: https://github.com/confluentinc/ksql/pull/6508
This is included in 0.14

**Comment by mikebin** [comment]
That's great - thanks @jzaralim

**Comment by mikebin** [comment]
Duplicate of https://github.com/confluentinc/ksql/issues/6506

**Comment by spena** [comment]
@mikebin I'll mark this as close if it is already fixed in 0.14.
